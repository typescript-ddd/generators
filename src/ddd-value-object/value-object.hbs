import { InvariantViolationError, ValueObject } from "@typescript-ddd/core";

/**
 * Represents {{noun className}} value.
 */
export class {{pascal className}} extends ValueObject<{{valueType}}> {
  /**
   * Initializes a new instance of {{noun className}} value.
   * @param {{lbrace}}{{valueType}}{{rbrace}} value - The value.
   */
  constructor(value: {{valueType}}) {
    super(value);
    if (!value) {
      throw new InvariantViolationError("value", "The {{natural className}} value cannot be null or undefined.");
    }
  }

  /**
   * @inheritdoc
   */
  protected selfEquals(other: {{pascal className}}): boolean {
    return this.value === other.value;
  }

  /**
   * Creates a new instance of {{noun className}} value.
   * @param value {{valueType}} - The value.
   * @returns {{lbrace}}{{pascal className}}{{rbrace}} - A new instance of {{noun className}} value.
   */
  public static create(value: {{valueType}}): {{pascal className}} {
    return new {{pascal className}}(value);
  }
}