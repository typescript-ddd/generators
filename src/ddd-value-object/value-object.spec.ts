import { generateValueObject } from "./value-object.generator";
import { ValueObjectTemplateData } from "./value-object.types";

describe("ValueObject", () => {

  it("should render template", () => {
    // Arrange
    const data: ValueObjectTemplateData = {
      className: "PhoneNumber",
      valueType: "string",
    };

    // Act
    const result = generateValueObject(data);

    // Assert
    expect(result).toMatchSnapshot();
  });
});
