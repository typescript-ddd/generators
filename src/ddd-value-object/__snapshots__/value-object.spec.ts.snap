// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ValueObject should render template 1`] = `
"import { InvariantViolationError, ValueObject } from "@typescript-ddd/core";

/**
 * Represents a phone number value.
 */
export class PhoneNumber extends ValueObject<string> {
  /**
   * Initializes a new instance of a phone number value.
   * @param {string} value - The value.
   */
  constructor(value: string) {
    super(value);
    if (!value) {
      throw new InvariantViolationError("value", "The phone number value cannot be null or undefined.");
    }
  }

  /**
   * @inheritdoc
   */
  protected selfEquals(other: PhoneNumber): boolean {
    return this.value === other.value;
  }

  /**
   * Creates a new instance of a phone number value.
   * @param value string - The value.
   * @returns {PhoneNumber} - A new instance of a phone number value.
   */
  public static create(value: string): PhoneNumber {
    return new PhoneNumber(value);
  }
}"
`;
